packages/blocks/src/store/selectors.js

```diff
@@ -95,6 +95,38 @@ export function getBlockVariations( state, blockName, scope ) {
 	} );
 }
 
+/**
+ * Returns the active block variation for a given block based on its attributes.
+ *
+ * @param {Object}                state      Data state.
+ * @param {string}                blockName  Name of block (example: “core/columns”).
+ * @param {Object}                attributes Block attributes used to determine active variation.
+ * @param {WPBlockVariationScope} [scope]    Block variation scope name.
+ *
+ * @return {(WPBlockVariation|void)} Active block variation.
+ */
+export function getActiveBlockVariation( state, blockName, attributes, scope ) {
+	const variations = getBlockVariations( state, blockName, scope );
+
+	const match = variations?.find( ( variation ) => {
+		if ( Array.isArray( variation.isActive ) ) {
+			const blockType = getBlockType( state, blockName );
+			const attributeKeys = Object.keys( blockType.attributes || {} );
+			return variation.isActive
+				.filter( ( attribute ) => attributeKeys.includes( attribute ) )
+				.every(
```

Comment:
We need to check if `.filter( ( attribute ) => attributeKeys.includes( attribute ) )` returns empty `array`.
If it does, in case of non existing attributes we must return `false` and bail early.

Now it will return `true` as this is how `every` works, resulting in returning the first declared block variation (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/every).

>Caution: Calling this method on an empty array will return true for any condition!

Reply:
Thanks! This will be obsolete once we validate variations at the time of block type registration.